### ISSUE TITLE
cibuildwheel 2.11.3/2.11.4 fails to build NumPy on 32-bit Windows

### ISSUE THREAD

**Issue by charris** [pypa/cibuildwheel#1389]
### Description

cibuildwheel 2.11.4 fails to build for cp39, cp310, and cp311 on 32 bit Windows, see https://github.com/numpy/numpy/pull/22921. A log can be found at https://github.com/numpy/numpy/actions/runs/3832441885/jobs/6522755840. 

The builds work for cibuildwheel 2.11.2.  There is also a pp38 segfault on manylinux2014, but that may be a PyPy fluke.

**Comment by henryiii** [comment]
```
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\runneradmin\\AppData\\Local\\Temp\\cibw-run-i4ooz105\\cp39-win32\\build\\venv\\libs\\python39.def'
```

(inside NumPy’s distutils)

We moved to using cmd instead of bash for the action, though I don’t know how that would affect this only for 32-bit. 3.8 and 3.9 haven’t updated, so it’s not a Python bump.

**Comment by joerick** [comment]
Setuptools updated in that time too, but numpy pins that in their build-system.requires.

The other thing that I notice is that virtualenv updated from `virtualenv==20.16.5` to `virtualenv==20.17.1`. Perhaps virtualenv isn't including that .defs file?

**Comment by henryiii** [comment]
Would it be possible to try 2.11.3, just to narrow it down one last little bit?

**Comment by charris** [comment]
> Would it be possible to try 2.11.3

Same problem: https://github.com/numpy/numpy/actions/runs/3839641498/jobs/6537708313.

I think the pp38 failure was also legitimate, but that takes about 90 minutes to build and test.

**Comment by henryiii** [comment]
There might be differences in configuration for `cmd` vs `bash`. I think it should still work, but numpy is somehow not detecting it correctly and not generating this file (this file seems to be written by numpy's distutils in https://github.com/numpy/numpy/blob/46142454ed4746ac991d4074d2943d892ede35d6/numpy/distutils/mingw32ccompiler.py#L324-L331).

Oh, `build_import_library` being called before `build_msvcr_library`... It's trying to write it in https://github.com/numpy/numpy/blob/46142454ed4746ac991d4074d2943d892ede35d6/numpy/distutils/mingw32ccompiler.py#L440-L446. The other paths (like 64-bit) look cleaner to me - the 32-bit one has extra handling for venvs, which I'm guessing may be the cause of this. I'm guessing it can't open this file for writing - maybe the "libs" folder doesn't exist?

**Comment by joerick** [comment]
These issues might be related? https://github.com/numpy/numpy/issues/20391 and the conversation in the linked PR https://github.com/numpy/numpy/pull/20383 .

**Comment by mattip** [comment]
Updating that one difference from a successful build is that the broken build is not using MSVC, rather mingw32, as the main compiler. Both runs are using the same version of the github runner with its builtin software.

**Comment by charris** [comment]
Closing. All issues except pp38 have been fixed on the NumPy side. The pp38 segfault should be fixed by #1372.
